src/os_pcap_nodelet.cpp

```diff
@@ -0,0 +1,247 @@
+            lidar_packets_processing_thread_active = false;
+            lidar_packets_processing_thread->join();
+        }
+    }
+
+    void on_lidar_packet_msg(const uint8_t* raw_lidar_packet) {
+        // copying the data from queue buffer into the message buffer
+        // this can be avoided by constructing an abstraction where
+        // OusterSensor has its own RingBuffer of PacketMsg but for
+        // now we are focusing on optimizing the code for OusterDriver
+        std::memcpy(lidar_packet.buf.data(), raw_lidar_packet,
+                    lidar_packet.buf.size());
+        lidar_packet_pub.publish(lidar_packet);
+    }
+
+    void on_imu_packet_msg(const uint8_t* raw_imu_packet) {
+        // copying the data from queue buffer into the message buffer
+        // this can be avoided by constructing an abstraction where
+        // OusterSensor has its own RingBuffer of PacketMsg but for
+        // now we are focusing on optimizing the code for OusterDriver
+        std::memcpy(imu_packet.buf.data(), raw_imu_packet,
+                    imu_packet.buf.size());
+        imu_packet_pub.publish(imu_packet);
+    }
+
+    void read_packets(PcapReader& pcap, const sensor::packet_format& pf) {
+        size_t payload_size = pcap.next_packet();
+        auto packet_info = pcap.current_info();
+        auto file_start = packet_info.timestamp;
+        auto last_update = file_start;
+        using namespace std::chrono_literals;
+        const auto UPDATE_PERIOD = duration_cast<microseconds>(1s / 3);
+
+        while (payload_size) {
+            auto start = high_resolution_clock::now();
+            if (packet_info.dst_port == info.config.udp_port_imu) {
+                imu_packets->write_overwrite(
+                    [this, &pcap, &pf, &packet_info](uint8_t* buffer) {
+                        std::memcpy(buffer, pcap.current_data(),
+                                    pf.imu_packet_size);
+                    });
+            } else if (packet_info.dst_port == info.config.udp_port_lidar) {
+                lidar_packets->write_overwrite(
+                    [this, &pcap, &pf, &packet_info](uint8_t* buffer) {
+                        std::memcpy(buffer, pcap.current_data(),
+                                    pf.lidar_packet_size);
+                    });
+            } else {
+                std::cout << "unknown packet" << std::endl;
+            }
+            auto prev_packet_ts = packet_info.timestamp;
+            payload_size = pcap.next_packet();
+            packet_info = pcap.current_info();
+            auto curr_packet_ts = packet_info.timestamp;
+            auto end = high_resolution_clock::now();
+            auto dt = (curr_packet_ts - prev_packet_ts) - (end - start);
+            std::this_thread::sleep_for(dt);  // pace packet generation
+
+            if (curr_packet_ts - last_update > UPDATE_PERIOD) {
+                last_update = curr_packet_ts;
+                std::cout << "\rtime passed: "
```

Comment:
Should this be a debug log message?

Reply:
Let me try if this can be a log message.

Reply:
The issue ROS log API adds a new line by default which wouldn't look nice for the way I designed it to be. also no benefit of directing this message to the log.